File: LogiOS/TrustarApp/Detect/DetectRestEvent.swift
//
//  DetectRestEvent.swift
//  TrustarApp
//
//  Created by DionSoftware on 26/03/2023.
//  Copyright © 2023 DionSoftware. All rights reserved.
//

import Foundation
import CoreLocation
import Combine

class DetectRestEvent {
    class MoveDetected: DetectRestEvent {
        var rest: RestDetected
        var endRealtimeNaos: Int64  //miliseconds
        
        init(rest: RestDetected, endRealtimeNaos: Int64) {
            self.rest = rest
            self.endRealtimeNaos = endRealtimeNaos
        }
        
        func elapsedRestMillis() -> Int64 {
            return (endRealtimeNaos - rest.startRestLocation.timestamp.milisecondsSince1970)
        }
        
        func restEndTimeBasedOsDate() -> Int64 {
            return dateBasedOnCurrentDate(realtimeNanos: self.endRealtimeNaos)
         }
    }
    
    class RestDetected: DetectRestEvent {
        var startRestLocation: CLLocation
        var speed: Double
        
        init(startRestLocation: CLLocation, speed: Double) {
            self.startRestLocation = startRestLocation
            self.speed = speed
        }
    }
    
    class Reset: DetectRestEvent {
        var elapsedMillis: Int64? = nil
        
        init (elapsedMillis: Int64?) {
            self.elapsedMillis = elapsedMillis
        }
    }
    
    init() {
        
    }
}

extension DetectRestEvent {
    func detectAfter(
        locationFlow: AnyPublisher<CLLocation, Never>,
        restDistance: Int,
        restSpeedMpMs: Double,
        restTimeInMillis: Int64,
        speedDetectSample: Int64
    ) -> AnyPublisher<DetectRestEvent, Never> {
        let behaviorProcessor = CurrentValueSubject<Optional<DetectRestEvent>, Never>(self)

        return behaviorProcessor
            .map({ e -> AnyPublisher<DetectRestEvent, Never> in
                if e is DetectRestEvent.RestDetected {
                    return locationFlow
                        .withPrevious()
                        .map({ (item) -> AnyPublisher<Optional<DetectRestEvent>, Never> in
                            let secondToLast = item.previous
                            let last = item.current
                            let eRD = e as! DetectRestEvent.RestDetected
                            if secondToLast != nil && last.distance(from: eRD.startRestLocation) > Double(restDistance) {
                                let t = last.elapsedMillis(eRD.startRestLocation)
                                if (t > restTimeInMillis) {
                                    
                                    DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + 0.1) {
                                        behaviorProcessor.send(MoveDetected(rest: eRD, endRealtimeNaos: secondToLast!.timestamp.milisecondsSince1970))
                                    }
                                    
                                    return Just(MoveDetected(rest: eRD, endRealtimeNaos: secondToLast!.timestamp.milisecondsSince1970)).eraseToAnyPublisher()
                                } else {
                                    DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + 0.1) {
                                        behaviorProcessor.send(Reset(elapsedMillis: t))
                                    }
                                    return Just(Reset(elapsedMillis: t)).eraseToAnyPublisher()
                                }
                            }
                            else {
                                return Just(nil).eraseToAnyPublisher()
                            }
                        })
                        .switchToLatest()
                        .compactMap({ $0 })
                        .eraseToAnyPublisher()
                } else {
                    return locationFlow
                    .locationSample(speedDetectSample: speedDetectSample)
                    .map( { (lastDropped, list)  -> AnyPublisher<Optional<DetectRestEvent>, Never> in
                            let last = list.last
                            if (lastDropped != nil && last != nil) {
                                let t = last!.elapsedMillis(lastDropped!)
                                let distance = last!.distance(from: lastDropped!)
                                let speed = Double(distance) / Double(t)
                                if (speed < restSpeedMpMs) {
                                    let secondToLast = list.secondToLast() ?? lastDropped
                                    
                                    DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + 0.1) {
                                        behaviorProcessor.send(RestDetected(startRestLocation: secondToLast!, speed: speed))
                                    }
                                    
                                    return Just(RestDetected(startRestLocation: secondToLast!, speed: speed)).eraseToAnyPublisher()
                                } else {
                                    return Just(nil).eraseToAnyPublisher()
                                }
                            } else {
                                return Just(nil).eraseToAnyPublisher()
                            }
                        }
                    )
                    .switchToLatest()
                    .compactMap({ $0 })
                    .eraseToAnyPublisher()
                }
            })
            .switchToLatest()
            .eraseToAnyPublisher()
    }
}

extension AnyPublisher<CLLocation, Never> {
    func locationSample(speedDetectSample: Int64) -> AnyPublisher<(CLLocation?, [CLLocation]), Never> {
        return keepHistoryUntil { current, item in
            current.elapsedMillis(item) >= speedDetectSample
        }.map { (drop, list) in
            return (drop.last, list)
        }.scan((nil,[])) { pre, cur in
            let lastDropped = cur.0 ?? pre.0
            return (lastDropped, cur.1)
        }
        .eraseToAnyPublisher()
    }
}

File: LogAndroid/TrustarApp/src/main/kotlin/jp/co/toukei/log/trustar/detect/DetectRestEvent.kt
package jp.co.toukei.log.trustar.detect

import android.location.Location
import io.reactivex.rxjava3.core.Flowable
import io.reactivex.rxjava3.processors.BehaviorProcessor
import jp.co.toukei.log.lib.dateBasedOnCurrentDate
import jp.co.toukei.log.lib.elapsedMillis
import jp.co.toukei.log.lib.observeOnIO
import jp.co.toukei.log.lib.secondToLast
import jp.co.toukei.log.lib.util.keepHistoryUntil
import jp.co.toukei.log.lib.withPrevious
import jp.co.toukei.log.trustar.detect.DetectRestEvent.MoveDetected
import jp.co.toukei.log.trustar.detect.DetectRestEvent.Reset
import jp.co.toukei.log.trustar.detect.DetectRestEvent.RestDetected

sealed class DetectRestEvent {

    class MoveDetected(
        val rest: RestDetected,
        private val endRealtimeNanos: Long,
    ) : DetectRestEvent() {

        fun elapsedRestMillis(): Long {
            return (endRealtimeNanos - rest.startRestLocation.elapsedRealtimeNanos) / 1000_000
        }

        fun restEndTimeBasedOsDate(): Long {
            return dateBasedOnCurrentDate(endRealtimeNanos)
        }
    }

    class RestDetected(
        val startRestLocation: Location,
        val speed: Double,
    ) : DetectRestEvent()

    class Reset(
        val elapsedMillis: Long? = null,
    ) : DetectRestEvent()
}

fun DetectRestEvent.detectAfter(
    locationFlow: Flowable<Location>,
    restDistance: Int,
    restSpeedMpMs: Double,
    restTimeInMillis: Long,
    speedDetectSample: Long,
): Flowable<DetectRestEvent> {
    val behaviorProcessor = BehaviorProcessor.createDefault(this)
    return behaviorProcessor
        .switchMap { e ->
            when (e) {
                is RestDetected -> locationFlow
                    .withPrevious()
                    .switchMap { (secondToLast, last) ->
                        if (secondToLast != null && last.distanceTo(e.startRestLocation) > restDistance) {
                            val t = last.elapsedMillis(e.startRestLocation)
                            Flowable.just(
                                if (t > restTimeInMillis) {
                                    MoveDetected(e, secondToLast.elapsedRealtimeNanos)
                                } else {
                                    Reset(t)
                                }
                            )
                        } else {
                            Flowable.empty()
                        }
                    }

                else -> locationFlow
                    .locationSample(speedDetectSample)
                    .switchMap { (lastDropped, list) ->
                        val last = list.lastOrNull()
                        if (lastDropped == null || last == null) {
                            Flowable.empty()
                        } else {
                            val t = last.elapsedMillis(lastDropped)
                            val distance = last.distanceTo(lastDropped)
                            val speed = distance.toDouble() / t
                            if (speed < restSpeedMpMs) {
                                val secondToLast = list.secondToLast() ?: lastDropped
                                Flowable.just(RestDetected(secondToLast, speed))
                            } else {
                                Flowable.empty()
                            }
                        }
                    }
            }
        }
        .doAfterNext(behaviorProcessor::onNext)
        .observeOnIO()
}

private fun Flowable<Location>.locationSample(speedDetectSample: Long): Flowable<Pair<Location?, List<Location>>> {
    return keepHistoryUntil { current, item ->
        current.elapsedMillis(item) >= speedDetectSample
    }.map { (drops, list) ->
        drops.lastOrNull() to list
    }.scan { (d0, _), (d, l) ->
        val lastDropped = d ?: d0
        lastDropped to l
    }
}
